NBitcoin/Miniscript/Satisfy.cs

```diff
@@ -0,0 +1,674 @@
+			{
+				result.AddRange(lSat);
+			}
+			else if (!isLOk && isROk)
+			{
+				var lDissat = Dissatisfy(l);
+				rSat.AddRange(lDissat);
+				result.AddRange(rSat);
+			}
+			else
+			{
+				var lDissat = Dissatisfy(l);
+				if (SatisfyCost(lSat) <= SatisfyCost(rSat) + SatisfyCost(lDissat))
+				{
+					result.AddRange(lSat);
+				}
+				else
+				{
+					rSat.AddRange(lDissat);
+					result.AddRange(rSat);
+				}
+			}
+			return true;
+		}
+
+		private bool SatisfyParallelOr(
+			AstElem l, AstElem r,
+			SignatureProvider signatureProvider,
+			PreimageProvider preimageProvider,
+			uint? age,
+			List<byte[]> result,
+			List<SatisfyError> errors
+			)
+		{
+			List<byte[]> lSat = new List<byte[]>();
+			List<byte[]> rSat = new List<byte[]>();
+			List<SatisfyError> leftE = new List<SatisfyError>();
+			List<SatisfyError> rightE = new List<SatisfyError>();
+			var isLOk = l.TrySatisfy(signatureProvider, preimageProvider, age, lSat, leftE);
+			var isROk = r.TrySatisfy(signatureProvider, preimageProvider, age, rSat, rightE);
+			if (!isLOk && !isROk)
+			{
+				leftE.AddRange(rightE);
+				errors.Add(new SatisfyError(SatisfyErrorCode.OrExpressionBothNotMet, this, leftE));
+				return false;
+			}
+
+			else if (isLOk && !isROk)
+			{
+				var rDissat = Dissatisfy(r);
+				rDissat.AddRange(lSat);
+				result.AddRange(rDissat);
+			}
+			else if (!isLOk && isROk)
+			{
+				var lDissat = Dissatisfy(l);
+				rSat.AddRange(lDissat);
+				result.AddRange(rSat);
+			}
+			else
+			{
+				var lDissat = Dissatisfy(l);
+				var rDissat = Dissatisfy(r);
+				if (SatisfyCost(lSat) + SatisfyCost(rDissat) <= SatisfyCost(rSat) + SatisfyCost(lDissat))
+				{
+					rDissat.AddRange(lSat);
+					result.AddRange(rDissat);
+				}
+				else
+				{
+					rSat.AddRange(lDissat);
+					result.AddRange(rSat);
+				}
+			}
+
+			return true;
+		}
+
+		private bool SatisfyHashEqual(
+			uint256 hash, PreimageProvider preimageProvider,
+			List<byte[]> result,
+			List<SatisfyError> errors
+			)
+		{
+			if (preimageProvider == null)
+			{
+				errors.Add(new SatisfyError(SatisfyErrorCode.NoPreimageProvider, this));
+				return false;
+			}
+			var preImage = preimageProvider(hash);
+			if (preImage == null)
+			{
+				errors.Add(new SatisfyError(SatisfyErrorCode.CanNotProvidePreimage, this));
+				return false;
+			}
+
+			result.Add(preImage.ToBytes());
+			return true;
+		}
+		private bool SatisfyCSV(
+			uint timelock, uint? age,
+			List<SatisfyError> errors
+			)
+		{
+			if (age == null)
+			{
+				errors.Add(new SatisfyError(SatisfyErrorCode.NoAgeProvided, this));
+				return false;
+			}
+			Sequence timelockS = timelock;
+			Sequence ageS = age.Value;
+			if (!ageS.IsRelativeLock)
+			{
+				errors.Add(new SatisfyError(SatisfyErrorCode.RelativeLockTimeDisabled, this));
+				return false;
+			}
+			if (ageS.LockType == SequenceLockType.Time)
+			{
+				errors.Add(new SatisfyError(SatisfyErrorCode.UnSupportedRelativeLockTimeType, this));
```

Comment:
Currently, I decided to support only block-height based lock for time (which is `OP_CSV` )
so in DSL it `"time(0) ~ time(65535)"` is allowed.

I don't know this will be standard among other Miniscript implementation. So it might change in the future.